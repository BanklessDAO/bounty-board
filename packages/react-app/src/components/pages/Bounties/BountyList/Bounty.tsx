import {
  AccordionButton,
  AccordionIcon,
  AccordionItem,
  AccordionPanel,
  Avatar,
  Badge,
  Box,
  Button,
  Flex,
  Grid,
  GridItem,
  Heading,
  Tag,
  TagLabel,
  Text,
} from '@chakra-ui/react'

const Status = ({ indication }: { indication: string }): JSX.Element => (
  <Tag my={0} size="lg" key="lg" variant="outline" colorScheme={indication}>
    <TagLabel>{indication.replace('-', ' ')}</TagLabel>
  </Tag>
)

const DiscordStub = ({ name }: { name: string }): JSX.Element => (
  <Flex my={2} align="center" gridGap={3}>
    <Avatar size="sm" src="https://bit.ly/broken-link" />
    <Text fontSize="md" color="steelblue">
      @{name}
    </Text>
  </Flex>
)

const Bounty = ({
  title,
  reward,
  status,
  guilds,
  description,
  criteria,
  author,
  claimed,
}: {
  title: string
  reward?: any
  status: string
  guilds: Array<string>
  description: string
  criteria: string
  author: string
  claimed?: string
}): JSX.Element => (
  <AccordionItem borderWidth={3} borderRadius={10} mb={3}>
    <AccordionButton pb={5}>
      <Flex flexWrap="wrap" width="100%" justifyContent="flex-end" ml="2">
        <Box
          width={{ base: '100%', md: '60%' }}
          pr={{ base: 7, md: 0 }}
          align="left"
          mt="4"
        >
          <Heading mb={4} size="md" flex={{ base: 1, md: 0 }}>
            {title}
          </Heading>
        </Box>
        <Box
          width={{ base: '50%', md: '30%' }}
          textAlign={{ base: 'left', md: 'right' }}
          mt={{ base: 0, md: 4 }}
          ml="auto"
          pr={7}
        >
          {reward && (
            <Heading mt={1} size="md">
              {reward.amt} {reward.token}
            </Heading>
          )}
        </Box>
        <Box pos="absolute" align="flex-end" textAlign="right" pt="2">
          <AccordionIcon />
        </Box>
        <Box mb={2} width="50%" textAlign={{ base: 'right', md: 'left' }}>
          <Status indication={status} />
        </Box>
        <Box
          width={{ base: '100%', md: '50%' }}
          textAlign={{ base: 'left', md: 'right' }}
          pr={7}
        >
          {guilds &&
            guilds.map((guild: string) => (
              <Badge
                key={guild}
                borderRadius="10"
                _notFirst={{ marginLeft: '10px' }}
                mb={2}
                py={1}
                px={3}
              >
                {guild}
              </Badge>
            ))}
        </Box>
      </Flex>
    </AccordionButton>
    <AccordionPanel mx={2}>
      <Grid gap={6}>
        <GridItem>
          <Heading size="sm">Description</Heading>
          <Text>{description}</Text>
        </GridItem>
        <GridItem>
          <Heading size="sm">Done Criteria</Heading>
          <Text>{criteria}</Text>
        </GridItem>
        <GridItem>
          <Heading size="sm">Requested By</Heading>
          <DiscordStub name={author} />
        </GridItem>
        <GridItem>
          <Heading size="sm">Claimed By</Heading>
          {claimed ? (
            <DiscordStub name={claimed} />
          ) : (
            <Button my={2} size="sm" colorScheme="green">
              <TagLabel>Claim It</TagLabel>
            </Button>
          )}
        </GridItem>
      </Grid>
    </AccordionPanel>
  </AccordionItem>
)

export default Bounty
