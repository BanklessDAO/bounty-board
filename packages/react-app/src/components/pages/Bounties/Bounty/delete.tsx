import RestrictedTo from '@app/components/global/Auth';
import bountyStatus from '@app/constants/bountyStatus';
import { useRoles } from '@app/hooks/useRoles';
import { useUser } from '@app/hooks/useUser';
import { BountyCollection } from '@app/models/Bounty';
import { Button, Box, useColorMode, Tooltip } from '@chakra-ui/react';
import { useMemo } from 'react';

const useShowDeleteBountyButton = (bounty: BountyCollection): boolean => {
	const { user } = useUser();
	const roles = useRoles();
	const show = useMemo(() => {
		const isOpen = bounty.status === bountyStatus.OPEN;
		const isOwnBounty = bounty.createdBy.discordId === user?.id;
		const isAdmin = roles.includes('admin');
		return (isAdmin || isOwnBounty) && isOpen;
	}, [bounty, roles, user]);
	return show;
};

export const BountyDeleteButton: React.FC<{ bounty: BountyCollection }> = ({
	bounty,
}) => {
	const { colorMode } = useColorMode();
	const show = useShowDeleteBountyButton(bounty);
	return (
		<>
			{show && (
				<RestrictedTo roles={['delete-own-bounty', 'admin']}>
					<Tooltip
						hasArrow
						label='Not implemented Yet'
						shouldWrapChildren
						mt="3"
					>
						<Box p={2}>
							<Button
								boxShadow={'md'}
								borderColor={colorMode === 'dark' ? 'primary.700' : 'primary.300'}
								size='md'
								width='200px'
							>
                             Delete this bounty
							</Button>
						</Box>
					</Tooltip>
				</RestrictedTo>
			)}
		</>
	);
};
