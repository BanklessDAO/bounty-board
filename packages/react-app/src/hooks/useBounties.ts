import { BountyCollection } from '../models/Bounty';
import useSWR from 'swr';
import axios from '../utils/AxiosUtils';
import { useContext, useMemo } from 'react';
import { CustomerContext } from '@app/context/CustomerContext';

// axios returns an AxiosResponse, with the payload in the `data` object
const axiosFetcher = (url: string) => axios.get(url).then(({ data }) => data.data);

interface BountySWRResponse {
	isLoading: boolean;
	isError: boolean;
}

interface useBountiesResponse extends BountySWRResponse {
	bounties?: BountyCollection[]
}

interface useBountyResponse extends BountySWRResponse {
	bounty?: BountyCollection
}

export function useBounties(url: string, ready = true): useBountiesResponse {
	/**
   * Wraps the SWR hook with additional loading state for spinners
   * Optioanlly pass a `ready` param to delay execution.
   */
	const { data, error } = useSWR<BountyCollection[], unknown>(ready ? url : null, axiosFetcher);
	return {
		bounties: data,
		isLoading: !error && !data,
		isError: !!error,
	};
}

export function useBounty(id?: string | string[]): useBountyResponse {
	const { customer } = useContext(CustomerContext);
	const { data, error } = useSWR<BountyCollection, unknown>(
		id
			? `/api/bounties/${id}`
			: null,
		axiosFetcher
	);
	const mismatchedBounty = useMemo(() => {
		return customer && data && customer.customerId !== data.customerId;
	}, [customer, data]);

	if (typeof id !== 'string' || mismatchedBounty) {
		return {
			bounty: undefined,
			isLoading: false,
			isError: true,
		};
	}

	return {
		bounty: data,
		isLoading: !error && !data,
		isError: !!error,
	};
}

export default useBounties;