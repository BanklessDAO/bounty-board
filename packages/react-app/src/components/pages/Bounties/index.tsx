import { Stack, Text, VStack, Button } from '@chakra-ui/react';
import { useColorMode } from '@chakra-ui/color-mode';
import BountyAccordion from './BountyAccordion';
import React, { useContext, useEffect, useState, useRef } from 'react';
import Filters from './Filters';
import useDebounce from '../../../hooks/useDebounce';
import { CustomerContext } from '../../../context/CustomerContext';
import { BANKLESS } from '../../../constants/Bankless';
import useBounties from '../../../hooks/useBounties';
import { BountyCollection } from '../../../models/Bounty';
import BountyPaginate from './Filters/bountyPaginate';
import { CSVLink } from 'react-csv';

export const PAGE_SIZE = 10;

type SetState<T extends any> = (arg: T) => void;

const maxPages = (bounties: BountyCollection | BountyCollection[] | undefined) => {
	if (!bounties || !Array.isArray(bounties)) return 0;
	const numFullPages = Math.floor(bounties.length / PAGE_SIZE);
	const hasExtraPage = bounties.length % PAGE_SIZE != 0;
	return hasExtraPage ? numFullPages + 1 : numFullPages;
};

const FilterResultPlaceholder = ({ message }: { message: string }): JSX.Element => (
	<Stack
		borderWidth={3}
		borderRadius={10}
		width={{ base: '95vw', lg: '700px' }}
		textAlign="center"
		direction="row"
		justify="center"
		align="center"
	>
		<Text	fontSize="lg">{ message }</Text>
	</Stack>
);

const SelectExport = (({ bounties, selectedBounties, setSelectedBounties }: {
	bounties: BountyCollection[] | undefined,
	selectedBounties: string[],
	setSelectedBounties: SetState<string[]>,
	}): JSX.Element => {

	const { colorMode } = useColorMode();
	const [csvData, setCsvData] = useState<BountyCollection[]>([]);
	const csvLink = useRef<CSVLink & HTMLAnchorElement & { link: HTMLAnchorElement }>(null);
	const csvHeaders = [
		{ label: 'ID', key: '_id' },
		{ label: 'Title', key: 'title' },
		{ label: 'Reward Amount', key: 'reward.amount' },
		{ label: 'Reward Currency', key: 'reward.currency' },
	  ];
	  

	const handleSelectAll = (): void => {
		if (bounties && selectedBounties.length < bounties.length) {
			setSelectedBounties(bounties.map(({ _id }) => _id));
		}
	};

	const handleClearAll = (): void => {
		setSelectedBounties([]);
	};

	const handleCSV = (): void => {
		if (bounties && csvLink.current) {
			const getCsvData = bounties.filter(item => selectedBounties.includes(item._id));
			setCsvData(getCsvData);
			console.log(getCsvData.length);
			// Force state to update before invoking the download
			setTimeout(() => {
				csvLink?.current?.link.click();
			});
		}
	};

	return (
		<Stack justify="space-between" direction="row" mt={1}>
			<Button
				p={2}
				disabled={!bounties || selectedBounties.length == bounties.length}
				size="xs"
				bg={colorMode === 'light' ? 'primary.300' : 'primary.700'}
				onClick={handleSelectAll}
			>
				Select All
			</Button>
			<Button
				p={2}
				disabled={!bounties || selectedBounties.length == 0}
				size="xs"
				bg={colorMode === 'light' ? 'primary.300' : 'primary.700'}
				onClick={handleClearAll}
			>
				Clear All
			</Button>
			<Button
				p={2}
				disabled={!bounties || selectedBounties.length == 0}
				size="xs"
				bg={colorMode === 'light' ? 'primary.300' : 'primary.700'}
				onClick={handleCSV}
			>
				Export
			</Button>
			<CSVLink
			  	data={csvData}
				headers={csvHeaders}
				filename='bounties.csv'
				className='hidden'
				ref={csvLink}
				target='_blank'
			/>
		</Stack>
	);
});

const Bounties = (): JSX.Element => {
	/* Bounties will fetch all data to start, unless a single bounty is requested */
	const [page, setPage] = useState(0);
	const [status, setStatus] = useState('Open');
	const [search, setSearch] = useState('');
	const [gte, setGte] = useState(0);
	// how to handle the lte === 0 case?
	const [lte, setLte] = useState(Infinity);
	const [sortBy, setSortBy] = useState('createdAt');
	const [sortAscending, setSortAscending] = useState(false);
	const [selectedBounties, setSelectedBounties] = useState<string[]>([]);
	const debounceSearch = useDebounce(search, 500, true);

	const { customer } = useContext(CustomerContext);
	const { customerId } = customer;

	let dynamicUrl = '/api/bounties';
	dynamicUrl += `?status=${status === '' ? 'All' : status}`;
	dynamicUrl += `&search=${debounceSearch}`;
	dynamicUrl += `&lte=${lte}`;
	dynamicUrl += `&gte=${gte}`;
	dynamicUrl += `&sortBy=${sortBy}`;
	dynamicUrl += `&asc=${sortAscending}`;
	dynamicUrl += `&customerId=${customerId ?? BANKLESS.customerId}`;

	useEffect(() => {
		setPage(0);
	}, [search, gte, lte, sortBy]);

	const { bounties, isLoading, isError } = useBounties(dynamicUrl);

	const paginatedBounties = bounties
		? bounties.slice(PAGE_SIZE * page, Math.min(bounties.length, PAGE_SIZE * (page + 1)))
		: [];

	const noResults = ((search || status) && bounties && paginatedBounties.length === 0);
	return (
		<>
			<Stack
				direction={{ base: 'column', lg: 'row' }}
				align="top"
				fontSize="sm"
				fontWeight="600"
				gridGap="4"
			>
				<VStack
				  gridGap="1px"
				>
					<Filters
						status={status} setStatus={setStatus}
						search={search} setSearch={setSearch}
						lte={lte} setLte={setLte}
						gte={gte} setGte={setGte}
						sortBy={sortBy} setSortBy={setSortBy}
						sortAscending={sortAscending} setSortAscending={setSortAscending}
					/>
					<SelectExport bounties={bounties} selectedBounties={selectedBounties} setSelectedBounties={setSelectedBounties}/>
				</VStack>
				{isError || noResults
					? <FilterResultPlaceholder message={'No Results'} />
					: isLoading
						? <FilterResultPlaceholder message={'Loading'} />
						: <BountyAccordion bounties={paginatedBounties} selectedBounties={selectedBounties} setSelectedBounties={setSelectedBounties}/>
				}
			</Stack>
			<BountyPaginate
				page={page}
				setPage={setPage}
				maxPages={maxPages(bounties)}
			/>
		</>
	);
};

export default Bounties;
