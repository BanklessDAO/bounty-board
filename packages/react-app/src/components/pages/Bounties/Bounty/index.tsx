import {
	AccordionButton,
	AccordionIcon,
	AccordionItem,
	AccordionPanel,
	Spacer,
	Box,
	Checkbox,
	Flex,
	Tooltip,
	Tag,
	TagLabel,
	Text,
	HStack,
	Stack,
	Heading,
	useDisclosure,
	Modal,
	ModalOverlay,
	ModalContent,
	ModalHeader,
	ModalFooter,
	ModalBody,
	ModalCloseButton,
	useColorModeValue,
} from '@chakra-ui/react';
import { useBounty } from '@app/hooks/useBounties';
import { BountyCollection } from '@app/models/Bounty';
import { BountyNotFound } from '@app/pages/[id]';
import { SkeletonCircle, SkeletonText } from '@chakra-ui/skeleton';
import BountyClaim from './claim';
import BountySubmit from './submit';
import { BountyEditButton } from './edit';
import { BountyDeleteButton } from './delete';

import UserAvatar from '@app/components/parts/UserAvatar';
import PAID_STATUS from '@app/constants/paidStatus';
import BOUNTY_STATUS from '@app/constants/bountyStatus';
import MiscUtils from '../../../../utils/miscUtils';
import DOMPurify from 'dompurify';
import { toHTML } from 'discord-markdown';
import ReactHtmlParser from 'react-html-parser';
import { BsThreeDots } from 'react-icons/bs';
import { useRouter } from 'next/router';
import { useMemo } from 'react';
import { useUser } from '@app/hooks/useUser';

type SetState<T extends any> = (arg: T) => void;

interface StatusProps {
	bounty: BountyCollection;
	withAvatar?: boolean;
}
const Status = ({ bounty, withAvatar }: StatusProps): JSX.Element => (
	<Tag my={0} size="lg" key="lg" variant="outline" colorScheme={bounty.status}>
		{withAvatar && (
			<Box ml={-1} mr={2}>
				<UserAvatar user={bounty.claimedBy} size="xs" />
			</Box>
		)}
		<TagLabel>{bounty.status.replace('-', ' ')}</TagLabel>
	</Tag>
);

const PaidStatus = ({ bounty }: { bounty: BountyCollection }): JSX.Element => (
	<Tag
		my={0}
		size="lg"
		key="lg"
		variant="outline"
		colorScheme={bounty.paidStatus ? bounty.paidStatus : PAID_STATUS.UNPAID}
	>
		<TagLabel>
			{bounty.paidStatus ? bounty.paidStatus : PAID_STATUS.UNPAID}
		</TagLabel>
	</Tag>
);

const BountySelect = ({
	selectedBounties,
	setSelectedBounties,
	bountyId,
}: {
	selectedBounties: string[];
	setSelectedBounties: SetState<string[]>;
	bountyId: string;
}): JSX.Element => {
	const updateSelectedBounties = (event: any): void => {
		const bId = event.target.value;
		if (!selectedBounties.includes(bId)) {
			setSelectedBounties([...selectedBounties, bId]);
		} else {
			setSelectedBounties(
				selectedBounties.filter((selectedBountyId) => {
					return selectedBountyId !== bId;
				})
			);
		}
	};

	return (
		<Checkbox
			size="sm"
			pt="2"
			px="2"
			value={bountyId}
			colorScheme="primary"
			onChange={updateSelectedBounties}
			isChecked={selectedBounties.includes(bountyId)}
		/>
	);
};

const DiscordStub = ({ name }: { name: string | undefined }): JSX.Element => (
	<Flex my={2} align="center" gridGap={3}>
		{name ? (
			<Text fontSize="md" color="steelblue">
				@{name}
			</Text>
		) : (
			'no one'
		)}
	</Flex>
);

const calculateReward = (_reward: BountyCollection['reward']): string => {
	return `${_reward.amount ?? 0} ${_reward.currency}`;
};

const Label = ({ keyword, noMargin = false }: { keyword: string; noMargin?: boolean }): JSX.Element => (
	<Tag
		size="sm"
		colorScheme="teal"
		variant="subtle"
		borderRadius="full"
		ml={noMargin ? 0 : 1.5}
	>
		<TagLabel>
			{keyword.replace('-', ' ')}
		</TagLabel>
	</Tag>
);

const BountyTags = ({
	tags,
	showAll = false,
}: {
	tags: { channelCategory: string, keywords: string[] };
	showAll: boolean;
}): JSX.Element => {
	let labels: string[] = [];

	if (tags.channelCategory) {
		labels = labels.concat(tags.channelCategory);
	} if (tags.keywords) {
		labels = labels.concat(tags.keywords);
	}

	return (
		showAll
			?
			<>
				{labels && labels.map((label, index) => <Label key={index} keyword={label} noMargin={index == 0} />)}
			</>
			:
			<>
				{labels && labels.slice(0, 3).map((label, index) => <Label key={index} keyword={label} noMargin={index == 0} />)}
				{
					labels.length > 3 &&
					<Tooltip
						hasArrow
						label={
							<Stack textAlign="center">
								{labels.slice(3).map((label, index) => <Text key={index}>{label}</Text>)}
							</Stack>
						}
						fontSize="md"
						closeDelay={500}
					>
						<Tag size="sm" colorScheme="teal" variant="subtle" borderRadius="full" ml={1.5}>
							<TagLabel><BsThreeDots /></TagLabel>
						</Tag>
					</Tooltip>

				}
			</>
	);
};

export const BountySummary = ({
	bounty,
}: {
	bounty: BountyCollection;
}): JSX.Element => {
	return (
		<Flex flexWrap="wrap" width="100%" justifyContent="flex-end" pl="2" pr="2">
			<Flex flexWrap="nowrap" width={{ base: '100%', md: '60%' }} mt="2">
				<Box>
					<UserAvatar user={bounty.createdBy} size="sm" />
				</Box>
				<Box pl="2" width="100%">
					<Heading mb={2} size="md" noOfLines={1} flex={{ base: 1, md: 0 }}>
						{bounty.title}
						{bounty.tags &&

							<Text as="span" ml={2} mt={2}>
								<BountyTags tags={bounty.tags} showAll={false} />
							</Text>
						}
					</Heading>
				</Box>
			</Flex>
			<Box
				width={{ base: '100%', md: '30%' }}
				textAlign={{ base: 'left', md: 'right' }}
				mt={{ base: 0, md: 2 }}
				ml="auto"
			>
				{bounty.reward && (
					<Heading mb={4} size="md">
						{calculateReward(bounty.reward)}
					</Heading>
				)}
			</Box>
			<Flex width="100%" justifyContent="space-between" alignItems="center">
				<Box mb={2}>
					for:{' '}
					{bounty.assignTo
						? '@' + bounty.assignTo.discordHandle
						: bounty.assign
							? '@' + bounty.assignedName
							: bounty.gateTo
								? '@' + bounty.gateTo[0].discordName
								: bounty.gate
									? '@' + bounty.gate[0]
									: 'anyone'}
				</Box>
				<Spacer />
				<Box mb={2} pr="2">
					{bounty.status && <Status bounty={bounty} withAvatar={true} />}
				</Box>
				<Box mb={2}>
					<PaidStatus bounty={bounty} />
				</Box>

			</Flex>
		</Flex>
	);
};

export const BountyHeader = ({
	bounty,
}: {
	bounty: BountyCollection;
}): JSX.Element => {
	return (
		<Flex flexWrap="nowrap" width={{ base: '100%', md: '60%' }}>
			<Box pl="2" width="100%">
				<Heading isTruncated mb={2} size="md" flex={{ base: 1, md: 0 }}>
					{bounty.title}
				</Heading>
			</Box>
		</Flex>
	);
};

export const BountyActions = ({
	bounty,
}: {
	bounty: BountyCollection;
	onCancel: () => void;
}): JSX.Element => {
	return (
		<Flex justifyContent={'flex-end'}>
			{bounty.status == BOUNTY_STATUS.DRAFT && <BountySubmit bounty={bounty} />}
			{bounty.status == BOUNTY_STATUS.OPEN && <BountyClaim bounty={bounty} />}
			{(bounty.status == BOUNTY_STATUS.DRAFT ||
				bounty.status == BOUNTY_STATUS.OPEN) && (
				<BountyEditButton bounty={bounty} />
			)}
			{(bounty.status == BOUNTY_STATUS.DRAFT ||
				bounty.status == BOUNTY_STATUS.OPEN) && (
				<BountyDeleteButton bounty={bounty} />
			)}

		</Flex>
	);
};

const BountyModal = ({
	bountyIn,
	isOpen,
	onClose,
}: {
	bountyIn: BountyCollection;
	isOpen: boolean;
	onClose: () => void;
}): JSX.Element => {

	// Keep the bounty data from changing what the modal is displaying if a rerender happens 
	const bounty = useMemo(() => bountyIn, [isOpen]);

	return (
		<Modal scrollBehavior={'inside'} isOpen={isOpen} onClose={onClose}>
			<ModalOverlay />
			<ModalContent maxW={'700px'} borderWidth={3}>
				<ModalHeader
					bg={useColorModeValue('gray.200', 'gray.600')}
					roundedTop="md"
				>
					<BountyHeader bounty={bounty} />
				</ModalHeader>
				<ModalCloseButton />
				<ModalBody>
					<BountyDetails bounty={bounty} />
				</ModalBody>
				<ModalFooter>
					<Stack direction='row' spacing={1} alignItems='center' alignContent='center' verticalAlign={'center'}>
						<BountyActions bounty={bounty} onCancel={onClose} />
					</Stack>

				</ModalFooter>
			</ModalContent>
		</Modal>

	);
};

const BountyDetails = ({
	bounty,
}: {
	bounty: BountyCollection;
}): JSX.Element => {
	const { user } = useUser();
	const dueAt = bounty.dueAt
		? MiscUtils.shortDate(new Date(bounty.dueAt))
		: 'unspecified';

	return (
		<Flex justifyItems={'center'} alignItems="center" flexWrap="wrap" width="100%" pl="2" pr="2">
			<Flex flexWrap="wrap" alignItems="center" width="100%" pb="3">
				<Box
					width="120px"
					display="inline-block"
					justifyContent="flex-end"
					alignItems="center"
					pr="2"
				>
					<Heading size="sm" m={0}>
						status
					</Heading>
				</Box>
				<Box pr="2">
					<Status bounty={bounty} />
				</Box>
				<Box pr="2">
					<PaidStatus bounty={bounty} />
				</Box>
				<Spacer />
				<Box>
					{bounty.reward && (
						<Heading size="md" m={0}>
							{calculateReward(bounty.reward)}
						</Heading>
					)}
				</Box>
			</Flex>
			<Flex flexWrap="wrap" alignItems="center" width="80%" pb="3">
				<Box
					width="120px"
					display="inline-block"
					justifyContent="flex-end"
					alignItems="center"
					pr="2"
				>
					<Heading size="sm" m={0}>
						created by
					</Heading>
				</Box>
				<Box pr="2">
					<UserAvatar user={bounty.createdBy} size="xs" />
				</Box>
				<Box width="120px" pr="2">
					<DiscordStub name={bounty.createdBy?.discordHandle} />
				</Box>
				<Spacer />
				<Box
					width="120px"
					display="inline-block"
					justifyContent="flex-end"
					alignItems="center"
					pr="2"
					ml="2"
				>
					<Heading size="sm" m={0}>
						created date
					</Heading>
				</Box>
				<Box>
					<Text as="span" fontSize="sm">
						{MiscUtils.shortDate(new Date(bounty.createdAt))}
					</Text>
				</Box>
			</Flex>
			<Flex flexWrap="wrap" alignItems="center" width="80%" pb="3">
				<Box
					width="120px"
					display="inline-block"
					justifyContent="flex-end"
					alignItems="center"
					pr="2"
				>
					<Heading size="sm" m={0}>
						claimed by
					</Heading>
				</Box>
				<Box pr="2">
					<UserAvatar user={bounty.claimedBy} size="xs" />
				</Box>
				<Box pr="2" width="120px">
					<DiscordStub name={bounty.claimedBy?.discordHandle} />
				</Box>
				<Spacer />
				<Box
					width="120px"
					display="inline-block"
					justifyContent="flex-end"
					alignItems="center"
					pr="2"
					ml="2"
				>
					<Heading size="sm" m={0}>
						due date
					</Heading>
				</Box>
				<Box>
					<Text as="span" fontSize="sm">
						{dueAt}
					</Text>
				</Box>
			</Flex>
			<Flex flexWrap="wrap" alignItems="center" width="100%" pb="3">
				<Box
					width="120px"
					display="inline-block"
					justifyContent="flex-end"
					alignItems="center"
					pr="2"
				>
					<Heading size="sm" m={0}>
						claimable by
					</Heading>
				</Box>
				<Box pr="2">
					<Text isTruncated={true} as="span" fontSize="sm">
						{bounty.assignTo
							? '@' + bounty.assignTo.discordHandle
							: bounty.assign
								? '@' + bounty.assignedName
								: bounty.gateTo
									? '@' + bounty.gateTo[0].discordName
									: bounty.gate
										? '@' + bounty.gate[0]
										: 'anyone'}
					</Text>
				</Box>
			</Flex>
			{bounty.tags &&
				<Flex flexWrap="wrap" alignItems="center" width="100%" pb="3">
					<Box
						width="120px"
						display="inline-block"
						justifyContent="flex-end"
						alignItems="center"
						pr="2"
					>
						<Heading size="sm" m={0}>
							tags
						</Heading>
					</Box>
					<Box pr="2">
						<BountyTags tags={bounty.tags} showAll={true} />
					</Box>
				</Flex>
			}
			<Heading mt="5" width="100%" size="md" mb="0">
				Description
			</Heading>
			<Text mt="2" fontSize="sm" ml="2">
				{ReactHtmlParser(
					DOMPurify.sanitize(toHTML(bounty.description || 'none'))
				)}
			</Text>
			<Heading mt="5" width="100%" size="md" pt="2" mb="0">
				Success Criteria
			</Heading>
			<Text mt="2" fontSize="sm" ml="2">
				{ReactHtmlParser(DOMPurify.sanitize(toHTML(bounty.criteria || 'none')))}
			</Text>
			{bounty.submissionNotes && ([bounty.createdBy.discordId, bounty.claimedBy?.discordId].includes(user?.id)) &&
				<>
					<Heading mt="5" width="100%" size="md" pt="2" mb="0">
						Submission Notes
					</Heading>
					<Text mt="2" fontSize="sm" ml="2">
						{ReactHtmlParser(DOMPurify.sanitize(toHTML(bounty.submissionNotes || 'none')))}
					</Text>
				</>
			}

		</Flex>
	);
};

export const BountyCard = ({
	bounty,
}: {
	bounty: BountyCollection;
}): JSX.Element => {
	const router = useRouter();

	const onCancel = () => {
		router.push('/');
	};
	return (
		<Box width={{ base: '95vw', lg: '700px' }}>
			<Box borderWidth={3} borderRadius={10} mb={3} p={4}>
				<Box pb={5}>
					<BountyHeader bounty={bounty} />
				</Box>
				<Box mx={2}>
					<BountyDetails bounty={bounty} />
				</Box>
				<Box>
					<BountyActions bounty={bounty} onCancel={onCancel} />
				</Box>
			</Box>
		</Box>
	);
};

export const AccordionBountyItem = ({
	bounty,
	selectedBounties,
	setSelectedBounties,
}: {
	bounty: BountyCollection;
	selectedBounties: string[];
	setSelectedBounties: SetState<string[]>;
}): JSX.Element => (
	<AccordionItem borderWidth={3} borderRadius={10} mb={3}>
		<BountySelect
			bountyId={bounty._id}
			selectedBounties={selectedBounties}
			setSelectedBounties={setSelectedBounties}
		/>
		<AccordionButton pb={5} pt={0}>
			<BountySummary bounty={bounty} />
			<Box
				pos="relative"
				textAlign="right"
				w={0}
				left={-4}
				top={{ base: -10, md: -5 }}
			>
				<AccordionIcon />
			</Box>
		</AccordionButton>
		<AccordionPanel mx={2}>
			<BountyDetails bounty={bounty} />
		</AccordionPanel>
	</AccordionItem>
);

export const BountyItem = ({
	initialBounty,
	selectedBounties,
	setSelectedBounties,
}: {
	initialBounty: BountyCollection;
	selectedBounties: string[];
	setSelectedBounties: SetState<string[]>;
}): JSX.Element => {
	const {
		isOpen: isBountyModalOpen,
		onOpen: onBountyModalOpen,
		onClose: onBountyModalClose,
	} = useDisclosure();
	const { bounty, isLoading } = useBounty(initialBounty._id);

	return (
		<Box w="100%" borderWidth={3} borderRadius={10} mb={1}>
			{isLoading ? (
				// if loading, show the loader
				<Box padding="6" boxShadow="lg">
					<SkeletonCircle size="10" />
					<SkeletonText mt="4" noOfLines={3} spacing="4" />
				</Box>
			) : bounty ? (
				<HStack>
					<BountySelect
						bountyId={bounty._id}
						selectedBounties={selectedBounties}
						setSelectedBounties={setSelectedBounties}
					/>
					<Box w="100%" pb={2} pt={0} onClick={onBountyModalOpen}>
						<BountySummary bounty={bounty} />
					</Box>
					<BountyModal
						isOpen={isBountyModalOpen}
						onClose={onBountyModalClose}
						bountyIn={bounty}
					/>
				</HStack>
			) : (
				<BountyNotFound />
			)}
		</Box>
	);
};
