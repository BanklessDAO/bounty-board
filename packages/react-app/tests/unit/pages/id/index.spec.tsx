import React from 'react';
import { mount, shallow } from 'enzyme';
import * as router from 'next/router';
import * as hooks from '../../../../src/hooks/useBounties';
import { NextRouter } from 'next/router';
import BountyPageLayout, { BountyLoader, BountyNotFound, BountyPage } from '../../../../src/pages/[id]';
import Layout from '../../../../src/components/global/SiteLayout';
import { BountyCollection } from '../../../../src/models/Bounty';
import { BountyCard } from '../../../../src/components/pages/Bounties/Bounty';

const CUSTOMERNAME = 'Test';

jest.mock('@/components/pages/Bounties', () => {
	return jest.fn(() => null);
});

jest.mock('@/components/global/SiteLayout', () => {
	return jest.fn(() => null);
});

describe('Testing the bounty id page', () => {
	beforeEach(() => {
		jest.spyOn(router, 'useRouter').mockImplementation(() => ({
			query: {
				id: '1',
			},
		}) as unknown as NextRouter);

		jest.spyOn(React, 'useContext').mockImplementation(() => ({
			customer: {
				customerName: CUSTOMERNAME,
			},
		}));
	});

	it('Renders with the correct customer name', () => {
		mount(<BountyPageLayout />);
		expect(Layout)
			.toHaveBeenCalledWith(
				expect.objectContaining({ title: `${CUSTOMERNAME} Bounty Board` })
				, {}
			);
	});

	it('Shows the loading page if loading', () => {
		jest.spyOn(hooks, 'useBounty').mockImplementation(() => ({
			isLoading: true,
			isError: false,
			bounty: undefined,
		}));
		const wrapper = shallow(<BountyPage />);
		expect(wrapper.containsMatchingElement(<BountyLoader />)).toEqual(true);
	});

	it('Shows the error page if error', () => {
		jest.spyOn(hooks, 'useBounty').mockImplementation(() => ({
			isLoading: false,
			isError: true,
			bounty: undefined,
		}));
		const wrapper = shallow(<BountyPage />);
		expect(wrapper.containsMatchingElement(<BountyNotFound />)).toEqual(true);
	});

	it('Shows no error if we dont have a bounty id', () => {
		jest.spyOn(router, 'useRouter').mockImplementation(() => ({
			query: {
				id: undefined,
			},
		}) as unknown as NextRouter);
		jest.spyOn(hooks, 'useBounty').mockImplementation(() => ({
			isLoading: false,
			isError: true,
			bounty: undefined,
		}));
		const wrapper = shallow(<BountyPage />);
		expect(wrapper.containsMatchingElement(<BountyNotFound />)).toEqual(false);
	});

	it('Otherwise renders the bounty', () => {
		jest.spyOn(hooks, 'useBounty').mockImplementation(() => ({
			isLoading: false,
			isError: false,
			bounty: { _id: 'test' } as unknown as BountyCollection,
		}));
		const wrapper = shallow(<BountyPage />);
		expect(wrapper.containsMatchingElement(
			<BountyCard { ...{} as BountyCollection } />
		)).toEqual(true);
	});
});