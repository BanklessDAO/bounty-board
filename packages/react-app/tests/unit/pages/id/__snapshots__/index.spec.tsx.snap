// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing the bounty id page Renders 1`] = `ShallowWrapper {}`;
